File: jacobi.jl
using LinearAlgebra
using Printf

function suma(A::AbstractMatrix, x::Array{Float64,1};)
    n = length(A)
    m = length(x)
    for i in 1:3
        for j in 1:3
            temp = 0
            temp = temp + A[i][j] * x[j]
        end
    end
end

function jacobi(A::AbstractMatrix, b::Array{Float64,1};
            delta::Number=10^-10,
            epsilon::Number=1/2*10^-4,
            maxiter::Integer=100)

    # n = length(b)
    x = zeros(3)
    # println(typeof(x))
    for k in 1:maxiter
        y = x
        for i in 1:3
            sum = b[i]
            diag = A[i][i]
            if abs(diag) < epsilon
                println("diagonal element too small")
            end
            for j in 1:3
                if j != i
                    sum = sum - A[i][j] * y[j]
                end
            end
            x[i] = sum / diag
        end
        println(k, x)
        if norm(x - y) < epsilon
            println(k, x)
        end
    end
    print(x)
end


A = [1.0 9 1; 4 1 -1; 1 -3 12]
b = [5.0; 3; 31;]

jacobi(A, b)

# n = length(A)
# println(n)
# x = zeros(3)

# println(suma(A, x))

File: Implementacja/jacobi.py
import numpy as np

class Jacobi:
    def __init__(self, A, b):
        self.maxiter = 100
        self.delta = 10**-10
        self.epsilon = 1/2*10**-4
        self.A = A
        self.b = b

    def calculate(self):
        x = np.array([0, 0, 0])
        n = len(self.b)
        for k in range(1, self.maxiter):
            y = x
            for i in range(n):
                sum = self.b[i]
                diag = self.A[i][i]
                if abs(diag) < self.delta:
                    print("za małe elementy na przekątnej")
                for j in range(n):
                    if j != i:
                        sum = sum - self.A[i][j] * y[j]
                x[i] = sum / diag
                print(k, x)
                if np.linalg.norm(x - y) < self.epsilon:
                    print(k, x)

Q = np.array([[1, 9, 1],
            [4, 1, -1],
            [1, -3, 12]])

b = np.array([5, 3, 31])

x = Jacobi(Q, b)
x.calculate()
